### PR TITLE
fix(vm-destroy): fix deleting of volumes while deleting the VM

### PR DESCRIPTION
**Pull request by qqcky**

### CODE REVIEW

File: src/app/reducers/vm/redux/vm.effects.ts [hunk 1]
```diff
@@ -362,7 +361,7 @@ export class VirtualMachinesEffects {
     });
 
   @Effect()
-  destroyVm$: Observable<Action> = this.actions$
+  destroyVm$ = this.actions$
```
Comment (HeyRoach) [inline src/app/reducers/vm/redux/vm.effects.ts]: Maybe you should add type?

File: src/app/reducers/vm/redux/vm.effects.ts [hunk 2]
```diff
@@ -374,19 +373,25 @@ export class VirtualMachinesEffects {
...
+                  new vmActions.ExpungeVmSuccess(action.payload),
+                  new volumeActions.DeleteVolumes(action.payload)
+                );
               } else {
                 this.jobsNotificationService.finish({
                   id: notificationId,
                   message: 'JOB_NOTIFICATIONS.VM.DESTROY_DONE'
                 });
-                return new vmActions.UpdateVM(vm);
+                return Observable.of<any>(
```
Comment (HeyRoach) [inline src/app/reducers/vm/redux/vm.effects.ts]: please check if we can get rid of any here

File: src/app/reducers/vm/redux/vm.effects.ts [hunk 3]
```diff
@@ -374,19 +373,25 @@ export class VirtualMachinesEffects {
...
           return this.vmService.command(action.payload, 'destroy', params)
-            .map(vm => {
+            .switchMap(vm => {
               if (params.expunge) {
                 this.jobsNotificationService.finish({
                   id: notificationId,
                   message: 'JOB_NOTIFICATIONS.VM.EXPUNGE_DONE'
                 });
-                return new vmActions.ExpungeVmSuccess(action.payload);
+                return Observable.of<any>(
```
Comment (HeyRoach) [inline src/app/reducers/vm/redux/vm.effects.ts]: please, check if we can get rid of any here

File: src/app/reducers/volumes/redux/volumes.effects.ts [hunk 4]
```diff
@@ -214,36 +216,49 @@ export class VolumesEffects {
...
+    .ofType(volumeActions.DELETE_VOLUMES)
+    .withLatestFrom(this.store.select(fromVolumes.selectAll))
+    .do(([action, volumes]: [volumeActions.DeleteVolumes, Array<Volume>]) => {
+      const vmVolumes = volumes
+        .filter((volume: Volume) => !volume.isRoot && volume.virtualMachineId === action.payload.id);
+
+      if (vmVolumes.length) {
+        this.dialogService.confirm({ message: 'DIALOG_MESSAGES.VM.CONFIRM_DRIVES_DELETION' })
+          .subscribe((res) => {
+            if (res) {
```
Comment (HeyRoach) [inline src/app/reducers/volumes/redux/volumes.effects.ts]: Maybe here we can use filter(res => !!res)

File: src/app/reducers/volumes/redux/volumes.effects.ts [hunk 5]
```diff
@@ -214,36 +216,49 @@ export class VolumesEffects {
       }).afterClosed();
     });
 
+  @Effect({ dispatch: false })
+  deleteVolumes$ = this.actions$
```
Comment (zolotyx) [inline src/app/reducers/volumes/redux/volumes.effects.ts]: Add typing

File: src/app/reducers/volumes/redux/volumes.effects.ts [hunk 6]
```diff
@@ -214,36 +216,49 @@ export class VolumesEffects {
...
+    .ofType(volumeActions.DELETE_VOLUMES)
+    .withLatestFrom(this.store.select(fromVolumes.selectAll))
+    .do(([action, volumes]: [volumeActions.DeleteVolumes, Array<Volume>]) => {
+      const vmVolumes = volumes
```
Comment (zolotyx) [inline src/app/reducers/volumes/redux/volumes.effects.ts]: use rxjs filter

File: src/app/shared/actions/volume-actions/volume-actions.container.ts [hunk 7]
```diff
@@ -33,7 +33,14 @@ export class VolumeActionsContainerComponent {
   }
 
   public onVolumeDelete(volume: Volume): void {
-    this.store.dispatch(new volumeActions.DeleteVolume(volume));
+    this.dialogService.confirm({
```
Comment (zolotyx) [inline src/app/shared/actions/volume-actions/volume-actions.container.ts]: why did you move this from effects?

**Comment by ksendart** [comment]
image
@zolotyx @qqcky  I rebuilt the project and got an error. Please, look through the code again

**Comment by ksendart** [comment]
ezgif com-video-to-gif
I tried to delete disk again, and the disk deletion still does not work. Also, you have conflicts, please, resolve them.
